server/storage/rethinkdb.go

```diff
@@ -0,0 +1,194 @@
+func NewRethinkDBStorage(dbName string, sess *gorethink.Session) RethinkDB {
+	return RethinkDB{
+		dbName: dbName,
+		sess:   sess,
+	}
+}
+
+// GetKey returns the cipher and public key for the given GUN and role.
+// If the GUN+role don't exist, returns an error.
+func (rdb RethinkDB) GetKey(gun, role string) (cipher string, public []byte, err error) {
+	var key RDBKey
+	res, err := gorethink.DB(rdb.dbName).Table(key.TableName()).Get(
+		RDBKey{
+			Gun:  gun,
+			Role: role,
+		}).Run(rdb.sess)
+	if err != nil {
+		return "", nil, err
+	}
+	defer res.Close()
+	err = res.One(&key)
+	return key.Cipher, key.Public, err
+}
+
+// SetKey sets the cipher and public key for the given GUN and role if
+// it doesn't already exist.  Otherwise an error is returned.
+func (rdb RethinkDB) SetKey(gun, role, cipher string, public []byte) error {
+	now := time.Now()
+	key := RDBKey{
+		Timing: rethinkdb.Timing{
+			CreatedAt: now,
+			UpdatedAt: now,
+		},
+		Gun:    gun,
+		Role:   role,
+		Cipher: cipher,
+		Public: public,
+	}
+	_, err := gorethink.DB(rdb.dbName).Table(key.TableName()).Insert(key).RunWrite(rdb.sess)
+	return err
+}
+
+// UpdateCurrent adds new metadata version for the given GUN if and only
+// if it's a new role, or the version is greater than the current version
+// for the role. Otherwise an error is returned.
+func (rdb RethinkDB) UpdateCurrent(gun string, update MetaUpdate) error {
+	now := time.Now()
+	checksum := sha256.Sum256(update.Data)
+	file := RDBTUFFile{
+		Timing: rethinkdb.Timing{
+			CreatedAt: now,
+			UpdatedAt: now,
+		},
+		Gun:     gun,
+		Role:    update.Role,
+		Version: update.Version,
+		Sha256:  hex.EncodeToString(checksum[:]),
+		Data:    update.Data,
+	}
+	_, err := gorethink.DB(rdb.dbName).Table(file.TableName()).Insert(
+		file,
+		gorethink.InsertOpts{
+			Conflict: "error", // default but explicit for clarity of intent
+		},
+	).RunWrite(rdb.sess)
+	if gorethink.IsConflictErr(err) {
+		return &ErrOldVersion{}
+	}
+	return err
+}
+
+// UpdateMany adds multiple new metadata for the given GUN. RethinkDB does
+// not support transactions, therefore we will attempt to insert the timestamp
+// first as this represents a published version of the repo. If this is successful,
+// we will insert the remaining roles (in any order). If any of those roles
+// errors on insert, we will do a best effort rollback, at a minimum attempting
+// to delete the timestamp so nobody pulls a broken repo.
+func (rdb RethinkDB) UpdateMany(gun string, updates []MetaUpdate) error {
+	for _, up := range updates {
+		if err := rdb.UpdateCurrent(gun, up); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// GetCurrent returns the modification date and data part of the metadata for
+// the latest version of the given GUN and role.  If there is no data for
+// the given GUN and role, an error is returned.
+func (rdb RethinkDB) GetCurrent(gun, role string) (created *time.Time, data []byte, err error) {
```

Comment:
We may need to continue to support `GetCurrent` for arbitrary roles in order to support older clients that don't support consistent downloads (I think Docker 1.10 does not, for instance).